src/utils.c

```diff
@@ -1,11 +1,18 @@
 #include "data.table.h"
 
+bool within_int32_repres(double x) {
+  // N.B. (int)2147483647.99 is not undefined behaviour since s 6.3.1.4 of the C
+  // standard states that behaviour is undefined only if the integral part of a
+  // finite value of standard floating type cannot be represented.
+  return R_FINITE(x) && x < 2147483648 && x > -2147483648;
```

Comment:
Remind me, `R_FINITE(NA_real_)` gives `FALSE` right? Just wondering for below in `firstNonInt`, how we can get `(int)x==NA_INTEGER` after the first two tests, is that last condition redundant?

Reply:
Is `data.table:::coerceAs(-2^31, NA_integer_)` intended to emit a warning? Under this pull request a warning will be emitted where it wasn't before, though this is inconsistent with base R. `as.integer(-2^31)`

Reply:
Yes, warning is expected

Reply:
As far as I can see, the only circumstance this comes up is for the peculiar:

```r
nafill(NA_integer_, fill = -2^31)
```

That is, setting fill to the double value that is the same as `NA_INTEGER`. I don't see any good reason that this would be done, intentionally or otherwise.